main.cpp

```diff
@@ -89,422 +89,222 @@ If you need to view an example, see: https://bitbucket.org/MatkatMusic/pfmcpptas
-{
-    if(intFunc)
+    Numeric& operator-=(const MyType other)
     {
-        intFunc(*intTypeHeap);
+        *numericHeap -= other;
+        return *this;
     }
-    return *this;
-}
-
-IntType& IntType::pow(int power)
-{
-    *intTypeHeap = powInternal(power);
-    return *this;
-}
-
-IntType& IntType::pow(const IntType& intRef)
-{
-    *intTypeHeap = powInternal(intRef);
-    return *this;   
-}
-
-IntType& IntType::pow(const DoubleType& doubleRef)
-{
-    *intTypeHeap = powInternal(static_cast<int>(doubleRef));
-    return *this;
-}
-
-int IntType::powInternal(int power)
-{
-    return static_cast<int>(std::pow(*intTypeHeap, power));
-}
-
-IntType& IntType::pow(const FloatType& floatRef)
-{
-    *intTypeHeap = (powInternal(static_cast<int>(floatRef)));
-    return *this;
-}
-
-//FloatType Implementation //////////////////////////////////////
-
-
-FloatType::FloatType(float floatValue)
-{
-    floatTypeHeap = new float(floatValue);
-}
-
-FloatType::~FloatType()
-{
-    delete floatTypeHeap;
-}
-
-FloatType& FloatType::operator+=( const float other) 
-{
-    *floatTypeHeap += other;
-    return *this;
-}
-
-FloatType& FloatType::operator-=(const float other)
-{
-    *floatTypeHeap -= other;
-    return *this;
-}
 
-FloatType& FloatType::operator/=( const float other) 
-{
-    *floatTypeHeap /= other;
-    return *this;
-}
-
-FloatType& FloatType::operator*=(const float other) 
-{
-    *floatTypeHeap *= other;
-    return *this;
-}
-
-FloatType& FloatType::apply(std::function<FloatType&(float&)> floatFunc)
-{
-    if(floatFunc)
+    Numeric& operator/=(const MyType other)
     {
-        return floatFunc(*floatTypeHeap);
+        // if your template type is an int
+        // if your parameter's type is also an int
+        //                 if your parameter is 0
+        //                         don't do the division
+        //         else if it's less than epsilon
+        //                 dont do the divison
+        // else if it's less than epsilon
+        //         warn about doing the division
+        if(std::is_same<int, MyType>::value)
+        {
+            if(std::is_same<decltype(other), int>::value)
+            {
+                if(other == 0.f)
+                {
+                    return *this;
+                }
+                else if(std::numeric_limits<MyType>::is_integer)
+                {
+                    return *this;
+                }
+                else
+                {
+                    std::cout<<"Warning division by zero";
+                }
+
+            } 
+        } 
+        return *this;
     }
-    return *this;
-}
 
-FloatType& FloatType::apply(void(*floatFunc)(float&))
-{
-    if(floatFunc)
+    Numeric& operator*=(const MyType other)
     {
-        floatFunc(*floatTypeHeap);
+        *numericHeap *= other;
+        return *this;
+    }
+    template<typename X>
+    Numeric& apply(X numericFunc)
+    {
+        numericFunc(numericHeap);
+        return *this;
```

Comment:
the `apply` function in the explicit template specialization should be templated.  
the regular template apply function is not templated.  See instruction 5
- [ ] revert to the previous version of `apply`.  the std::function version